table/tables/cache.go

```diff
@@ -30,74 +35,158 @@ var _ table.CachedTable = &cachedTable{}
+	}
+	ret := &cachedTable{
 		TableCommon: *tbl,
-	}, nil
+		handle:      &mockStateRemoteHandle{mockStateRemote.Ch},
+	}
+	return ret, nil
 }
 
-func (c *cachedTable) loadDataFromOriginalTable(ctx sessionctx.Context) error {
+// Close TODO only for skip go leap. and will remove in the following
+func Close() {
+	_, ok := <-mockStateRemote.Ch
+	if ok {
+		close(mockStateRemote.Ch)
+	}
+}
+func (c *cachedTable) loadDataFromOriginalTable(ctx sessionctx.Context) (kv.MemBuffer, error) {
 	prefix := tablecodec.GenTablePrefix(c.tableID)
 	txn, err := ctx.Txn(true)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	buffTxn, err := ctx.GetStore().BeginWithOption(tikv.DefaultStartTSOption().SetStartTS(0))
 	if err != nil {
-		return err
+		return nil, err
 	}
+
 	buffer := buffTxn.GetMemBuffer()
 	it, err := txn.Iter(prefix, prefix.PrefixNext())
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer it.Close()
 	if !it.Valid() {
-		return nil
+		return nil, nil
 	}
 	for it.Valid() && it.Key().HasPrefix(prefix) {
 		value := it.Value()
 		err = buffer.Set(it.Key(), value)
 		if err != nil {
-			return err
+			return nil, err
 		}
 		err = it.Next()
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
-	c.mu.Lock()
-	c.MemBuffer = buffer
-	c.mu.Unlock()
-	return nil
+
+	return buffer, nil
 }
 
 func (c *cachedTable) UpdateLockForRead(ctx sessionctx.Context, ts uint64) error {
-	// Now only the data is re-load here, and the lock information is not updated. any read-lock information update will in the next pr.
-	err := c.loadDataFromOriginalTable(ctx)
+	// Load data from original table and the update lock information.
+	tid := c.Meta().ID
+	lease := leaseFromTS(ts)
+	succ, err := c.handle.LockForRead(tid, ts, lease)
 	if err != nil {
-		return fmt.Errorf("reload data error")
+		return errors.Trace(err)
+	}
+	if succ {
+		mb, err := c.loadDataFromOriginalTable(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		c.cacheData.Store(&table.CacheData{
+			Lease:     lease,
+			MemBuffer: mb,
+		})
 	}
+	// Current status is not suitable to cache.
 	return nil
 }
+
+// AddRecord implements the AddRecord method for the table.Table interface.
+func (c *cachedTable) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ...table.AddRecordOption) (recordID kv.Handle, err error) {
+	txn, err := ctx.Txn(true)
+	if err != nil {
+		return nil, err
+	}
+	now := txn.StartTS()
+	err = c.handle.LockForWrite(c.Meta().ID, now, leaseFromTS(now))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return c.TableCommon.AddRecord(ctx, r, opts...)
+
+}
+
+// UpdateRecord implements table.Table
+func (c *cachedTable) UpdateRecord(ctx context.Context, sctx sessionctx.Context, h kv.Handle, oldData, newData []types.Datum, touched []bool) error {
+	txn, err := sctx.Txn(true)
+	if err != nil {
+		return err
+	}
+	now := txn.StartTS()
+	err = c.handle.LockForWrite(c.Meta().ID, now, leaseFromTS(now))
```

Comment:
I think it still has some risk when commit phase is slow that causes commitTs > lease

Reply:
Good point, the transaction should update write lock lease to make sure write lock lease  > commitTs

Renew lease will be in another PR